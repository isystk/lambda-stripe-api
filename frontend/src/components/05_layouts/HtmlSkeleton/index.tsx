import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren & {
  showSideMenu?: boolean
}
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  showSideMenu,
  main,
  title,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
      {/* Tailwind UI で採用されているカスタムフォント */}
      <link rel="stylesheet" href="https://rsms.me/inter/inter.css" />
    </Head>
    <Context.Provider value={main}>{children}</Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()

  if (!main) return <></>

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : child
  )
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    main,
    title,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
