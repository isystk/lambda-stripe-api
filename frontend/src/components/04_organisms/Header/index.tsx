import React, { FC, useContext } from 'react'
import { ContainerProps, WithChildren } from 'types'
import { connect } from '@/components/hoc'
import Logo from '@/components/01_atoms/Logo'
import Hamburger from '@/components/01_atoms/Hamburger'
import * as styles from './styles'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'
import { Url } from '@/constants/url'

/** HeaderProps Props */
export type HeaderProps = WithChildren & { isMenuOpen; isPcSize; setMenuOpen }
/** Presenter Props */
export type PresenterProps = HeaderProps

/** Presenter Component */
const HeaderPresenter: FC<PresenterProps> = ({
  main,
  isMenuOpen,
  setMenuOpen,
}) => (
  <>
    <header
      className={`${styles.header} flex justify-between items-center px-4 py-4 sm:px-8 bg-gray-100 w-full`}
    >
      <Logo />
      <div className="flex pr-3 z-50 md:hidden">
        <Hamburger isMenuOpen={isMenuOpen} setMenuOpen={setMenuOpen} />
      </div>
      <nav className="hidden sm:block">
        <ul className="flex gap-6">
          <li>
            <a href={Url.TOP}>ホーム</a>
          </li>
          <li>
            <a href="#">機能</a>
          </li>
          <li>
            <a href="#">料金</a>
          </li>
          <li>
            <a href="https://twitter.com/ise0615" target="_blank">お問い合わせ</a>
          </li>
        </ul>
      </nav>
    </header>
  </>
)

/** Container Component */
const HeaderContainer: React.FC<
  ContainerProps<HeaderProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>
  return presenter({
    children,
    main,
    ...props,
  })
}

export default connect<HeaderProps, PresenterProps>(
  'Header',
  HeaderPresenter,
  HeaderContainer
)
