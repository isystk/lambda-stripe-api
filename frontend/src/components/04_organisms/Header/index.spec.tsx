import React from 'react'
import renderer from 'react-test-renderer'
import Header from './index'

import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'

describe('Header', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))

    const component = renderer.create(
      <Context.Provider value={main}>
        <Header
          isMenuOpen={false}
          setMenuOpen={(isOpen) => console.log(isOpen)}
        />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
