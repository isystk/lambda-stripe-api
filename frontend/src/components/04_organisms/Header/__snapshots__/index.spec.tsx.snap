// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Match Snapshot 1`] = `
<header
  className="css-0 flex justify-between items-center px-4 py-4 sm:px-8 bg-gray-100 w-full"
>
  <a
    className="css-0 w-20 md:w-40"
    href="/"
  >
    <img
      src="/images/logo.png"
    />
  </a>
  <div
    className="flex pr-3 z-50 md:hidden"
  >
    <button
      className="space-y-2 css-jgxxba"
      onClick={[Function]}
    >
      <div
        className="w-6 h-0.5 bg-black"
      />
      <div
        className="w-6 h-0.5 bg-black"
      />
      <div
        className="w-6 h-0.5 bg-black"
      />
    </button>
  </div>
  <nav
    className="hidden sm:block"
  >
    <ul
      className="flex gap-6"
    >
      <li>
        <a
          href="/"
        >
          ホーム
        </a>
      </li>
      <li>
        <a
          href="/product/prod_NpvV9ohJIlgElI/payment"
        >
          料金
        </a>
      </li>
      <li>
        <a
          href="https://twitter.com/ise0615"
          rel="noreferrer"
          target="_blank"
        >
          お問い合わせ
        </a>
      </li>
    </ul>
  </nav>
</header>
`;
